### ISSUE TITLE
Bug: Many datetime format errors since updated to MISP 2.4.170

### ISSUE THREAD

**Issue by obert01** [MISP/MISP#9019]
### Expected behavior

No SQL error. 

### Actual behavior

Since the update to MISP 2.4.170 this morning, we encounter various "Internal error"  (code 500) when either using the API or the web interface. According to the logs, these errors are due to malformed datetime in SQL queries (e.g. adding a log entry, adding an event_blocklists entry, ...). 

In the "Diagnostics" page, I have verified that my current database schema matches the expected one. I have fixed all differences, including minor ones. 

It is a critical issue for us, since e.g. sync operations (pull / push) don't work anymore. 

Curiously, on a testing server where MISP has also been updated, the issue does not occur. The only difference I see between both is the MariaDB version:

- The working server runs MariaDB 10.3 on Ubuntu 20.04
- The non-working server runs MariaDB 10.6 on Ubuntu 22.04. 

Is it the explanation? 

Please refer to the logs below for the details. 

### Steps to reproduce

For example:

1. In MISP settings, enable MISP.log_auth if not already done. 
2. Perform any HTTP request to the API, for example /events/restSearch. 
3. You should obtain an error 500 - Internal Error

### Version

2.4.170

### Operating System

Ubuntu

### Operating System version

22.04

### PHP version

7.4

### Relevant log output

```shell
2023-04-14 11:48:11 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1292 Incorrect datetime value: '1681465691' for column `misp`.`logs`.`created` at row 1
Request URL: /events/restSearch/
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1132): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1942): DboSource->create()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave()
#5 /var/www/MISP/app/Controller/AppController.php(437): Model->save()
#6 /var/www/MISP/app/Controller/AppController.php(233): AppController->__loginByAuthKey()
#7 /var/www/MISP/app/Controller/EventsController.php(68): AppController->beforeFilter()
#8 /var/www/MISP/app/Lib/Tools/BetterCakeEventManager.php(23): EventsController->beforeFilter()
#9 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(682): BetterCakeEventManager->dispatch()
#10 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(189): Controller->startupProcess()
#11 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#12 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch()
#13 {main}

2023-04-14 13:30:14 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1292 Incorrect datetime value: '1681471814' for column `misp`.`event_blocklists`.`created` at row 1
Request URL: /events/730d4504-87c1-4d97-a1a8-538a9452b1bd
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1132): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1942): DboSource->create()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave()
#5 /var/www/MISP/app/Model/Event.php(324): Model->save()
#6 /var/www/MISP/app/Lib/Tools/BetterCakeEventManager.php(21): Event->beforeDelete()
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(2694): BetterCakeEventManager->dispatch()
#8 /var/www/MISP/app/Model/Event.php(1337): Model->delete()
#9 /var/www/MISP/app/Controller/EventsController.php(2833): Event->quickDelete()
#10 [internal function]: EventsController->delete()
#11 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs()
#12 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction()
#13 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#14 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch()
#15 {main}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by righel** [comment]
Hello @obert01 

Could you tell me which data type are these columns `misp`.`logs`.`created` table and  `misp`.`event_blocklists`.`created` in both of your databases (mariadb10.3 and mariadb10.6) ?

I was not able to reproduce the issue.

**Comment by obert01** [comment]
On both servers, they are declared as follows:
```
`created` datetime NOT NULL
```

**Comment by obert01** [comment]
Because the only difference I can figure between the two servers is the MariaDB version, I will try to update my dev server to MariaDB 10.6 and see if it makes the problem appear. 

For the moment, I have just run "git checkout tags/v2.4.169" on my production system and the problem disappeared. ^^

**Comment by obert01** [comment]
Hello,

I can't reproduce the issue on my test server which I have just upgraded to Ubuntu 22.04 like the prod system. 

The essential difference between both servers is that the prod server (which has issues) has been installed in 2016, and has been subject to several big in-place system upgrades since then. The dev system (which has no issue) has been installed more recently.

**Comment by obert01** [comment]
The issue is likely not due to MISP code.

**Comment by righel** [comment]
good to hear!

**Comment by muhyuddin** [comment]
Facing same issue on 2.4.170. Whenever trying to run connection test.

```
2023-05-02 22:10:01 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect string value: '\x8B\x08\x00\x00\x00\x00...' for column `misp`.`logs`.`change` at row 1
Request URL: /servers/postTest
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1132): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1942): DboSource->create()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave()
#5 /var/www/MISP/app/Model/Log.php(257): Model->save()
#6 /var/www/MISP/app/Controller/AppController.php(726): Log->createLogEntry()
#7 /var/www/MISP/app/Controller/AppController.php(310): AppController->__accessMonitor()
#8 /var/www/MISP/app/Controller/ServersController.php(41): AppController->beforeFilter()
#9 /var/www/MISP/app/Lib/Tools/BetterCakeEventManager.php(23): ServersController->beforeFilter()
#10 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(682): BetterCakeEventManager->dispatch()
#11 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(189): Controller->startupProcess()
#12 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#13 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch()
#14 {main}
```


Any suggestion to fix this issue ?

**Comment by iglocska** [comment]
After loads of frustrations and heroic debugging efforts from https://github.com/3c7 we now know what causes it.

MariaDB tables created in older versions of MariaDB used a different internal datetime field mechanism. When you updated your MariaDB to a more recent version, the tables did not get updated, and MariaDB handles this by trying to "warn" the users that the field is still in the old datetime format.

What this looks like in practice when you run a describe?

```
MariaDB [misp]> describe event_blocklists;
+------------+----------------------------+------+-----+---------+----------------+
| Field      | Type                       | Null | Key | Default | Extra          |
+------------+----------------------------+------+-----+---------+----------------+
| id         | int(11)                    | NO   | PRI | NULL    | auto_increment |
| event_uuid | varchar(40)                | NO   | UNI | NULL    |                |
| created    | datetime /* mariadb-5.3 */ | YES  |     | NULL    |                |
| event_info | text                       | NO   |     | NULL    |                |
| comment    | text                       | YES  |     | NULL    |                |
| event_orgc | varchar(255)               | NO   | MUL | NULL    |                |
+------------+----------------------------+------+-----+---------+----------------+
```

For some reason they figured that barfing this directly back in the type column would be helpful - sadly in our case it's quite the opposite. CakePHP does column introspection for each model, matching each column type to formatters for data entry. "datetime" gets mapped to a handler that will pass back datetime strings. "datetime /* mariadb-5.3 */" doesn't match anything and thus it seems to fall back to integers.

tl;dr:

```
ALTER TABLE `logs` MODIFY `created` datetime;
ALTER TABLE `event_blocklists` MODIFY `created` datetime;
```

will convert the datetime fields to the correct format and get rid of the dodgy name in the describe output, solving the issue.

We'll be implementing a diagnostic tool for this in the future, for now this should be a quick fix though.

**Comment by RR19MR** [comment]
Everything worked after the updates, but one day I got same error:
```
2025-01-03 08:17:41 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1292 Incorrect datetime value: '1735892261' for column `misp`.`logs`.`created` at row 1
Request URL: /events/edit/dc608d28-a388-49c2-b5f3-60fb53799043/metadata:1
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1132): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1943): DboSource->create()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1761): Model->_doSave()
#5 /var/www/MISP/app/Model/Log.php(322): Model->save()
#6 /var/www/MISP/app/Model/Event.php(4349): Log->createLogEntry()
#7 /var/www/MISP/app/Controller/EventsController.php(2854): Event->_edit()
#8 [internal function]: EventsController->edit()
#9 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(500): ReflectionMethod->invokeArgs()
```

Do you know what solutions could be available? The solutions provided at https://github.com/MISP/MISP/issues/9019 does not helping.

MISP version: 2.5.4

Describe logs
--------------

```
+-------------+--------------+------+-----+---------+----------------+
| Field       | Type         | Null | Key | Default | Extra          |
+-------------+--------------+------+-----+---------+----------------+
| id          | int(11)      | NO   | PRI | NULL    | auto_increment |
| title       | text         | YES  |     | NULL    |                |
| created     | datetime     | YES  |     | NULL    |                |
| model       | varchar(80)  | NO   |     | NULL    |                |
| model_id    | int(11)      | NO   |     | NULL    |                |
| action      | varchar(20)  | NO   |     | NULL    |                |
| user_id     | int(11)      | NO   |     | NULL    |                |
| change      | text         | YES  |     | NULL    |                |
| email       | varchar(255) | NO   |     |         |                |
| org         | varchar(255) | NO   |     |         |                |
| description | text         | YES  |     | NULL    |                |
| ip          | varchar(45)  | NO   |     |         |                |
+-------------+--------------+------+-----+---------+----------------+
```

**Comment by ofirzeiri** [comment]
> Facing same issue on 2.4.170. Whenever trying to run connection test.
> 
> ```
> 2023-05-02 22:10:01 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect string value: '\x8B\x08\x00\x00\x00\x00...' for column `misp`.`logs`.`change` at row 1
> Request URL: /servers/postTest
> Stack Trace:
> #0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
> #1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
> #2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1132): DboSource->execute()
> #3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1942): DboSource->create()
> #4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave()
> #5 /var/www/MISP/app/Model/Log.php(257): Model->save()
> #6 /var/www/MISP/app/Controller/AppController.php(726): Log->createLogEntry()
> #7 /var/www/MISP/app/Controller/AppController.php(310): AppController->__accessMonitor()
> #8 /var/www/MISP/app/Controller/ServersController.php(41): AppController->beforeFilter()
> #9 /var/www/MISP/app/Lib/Tools/BetterCakeEventManager.php(23): ServersController->beforeFilter()
> #10 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(682): BetterCakeEventManager->dispatch()
> #11 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(189): Controller->startupProcess()
> #12 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
> #13 /var/www/MISP/app/webroot/index.php(99): Dispatcher->dispatch()
> #14 {main}
> ```
> 
> Any suggestion to fix this issue ?

How did you solve this?
